### PR TITLE
Termination cause mapping

### PR DESCRIPTION
**Pull request by vkatsuba**

### CODE REVIEW

**Review by ebengt — APPROVED** [review]

File: src/ergw_aaa_session.erl [hunk 1]
```diff
@@ -474,3 +475,13 @@ get_handler_state(_, undefined) ->
     undefined;
 get_handler_state(Handler, State) ->
     Handler:get_state_atom(State).
+
+termination_cause_mapping(#{'Termination-Cause' := Cause} = Session, #{termination_cause_mapping := Config}) when is_atom(Cause) ->
+    case proplists:get_value(Cause, Config) of
```
[the remark this answers is not in the document]
Comment (vkatsuba) [reply src/ergw_aaa_session.erl:480]: `termination_cause_mapping` has configuration as list, see `DEFAULT_TERMINATION_CAUSE_MAP` macros. We can change it from `list` to `maps` but it looks like other configurations have a list view too.

File: include/ergw_aaa.hrl [hunk 2]
```diff
@@ -0,0 +1,10 @@
+%% Default reason mapping is not ideal but made to preserve the existing behaviour
+-define(DEFAULT_TERMINATION_CAUSE_MAP, [
```
[the remark this answers is not in the document]
Comment (vkatsuba) [reply include/ergw_aaa.hrl:2]: `map` this is short for `mapping`, but if this name is confusing I will add `PPING` to end of macros name

File: src/ergw_aaa_session.erl (hunk 1 above)
[the remark this answers is not in the document]
Comment (ebengt) [reply src/ergw_aaa_session.erl:480]: See ergw_aaa_config:load_config/0 
It changes config from proplist to map.
So, proplist in config file, but map when used.
Generally.

File: src/ergw_aaa_diameter.erl [hunk 3]
```diff
@@ -150,6 +152,17 @@ validate_transport(Opt, Value) ->
 validate_transport_error(Opt, Value) ->
     throw({error, {options, {Opt, Value}}}).
 
+validate_handler(Opts) ->
+    ergw_aaa_config:validate_options(fun validate_option/2, Opts, ?DefaultOptions, map).
+
+validate_option(termination_cause_mapping = Opt, Value) when is_list(Value) ->
+    case [{N, V} || {N, V} <- Value, is_atom(N), is_integer(V)] of
```
Comment (RoadRunnr) [inline src/ergw_aaa_diameter.erl:159]: and adding to the above comment, this should use `ergw_aaa_config:validate_options`

**Review by RoadRunnr — CHANGES_REQUESTED** [review]

File: src/ergw_aaa_diameter.erl (hunk 3 above)
Comment (vkatsuba) [reply src/ergw_aaa_diameter.erl:159]: But I already use `ergw_aaa_config:validate_options` in `validate_handler/1` and `validate_option` this is private function. Can you provide example what you expected here?

File: src/ergw_aaa_diameter.erl [hunk 4]
```diff
@@ -10,8 +10,7 @@
 -compile({parse_transform, cut}).
 
 %% API
--export([validate_function/1,
-	 initialize_function/2]).
+-export([validate_function/1, validate_handler/1, initialize_function/2]).
```
Comment (RoadRunnr) [inline src/ergw_aaa_diameter.erl:13]: ergw_aaa_diameter is not a handler, so adding that here make no sense.

File: src/ergw_aaa_gx.erl [hunk 5]
```diff
@@ -253,6 +255,11 @@ validate_option(answer_if_timeout, Value) when is_atom(Value) ->
     Value;
 validate_option(avp_filter, Value) when is_list(Value) ->
     Value;
+validate_option(termination_cause_mapping = Opt, Value) when is_list(Value) ->
+    case [{N, V} || {N, V} <- Value, is_atom(N), is_integer(V)] of
+        Value -> Value;
+        _ -> throw({error, {options, {Opt, Value}}})
+    end;
```
Comment (RoadRunnr) [inline src/ergw_aaa_gx.erl:262]:
```suggestion
validate_option(termination_cause_mapping, Value)
  when is_list(Value); is_map(Value) ->
    ergw_aaa_diameter:validate_term_cause_mapping(Value);
```

File: src/ergw_aaa_diameter.erl [hunk 6]
```diff
@@ -150,6 +152,17 @@ validate_transport(Opt, Value) ->
 validate_transport_error(Opt, Value) ->
     throw({error, {options, {Opt, Value}}}).
 
+validate_handler(Opts) ->
```
Comment (RoadRunnr) [inline src/ergw_aaa_diameter.erl:155]:
```suggestion
validate_term_cause_mapping(Opt, Value)
  when is_atom(Opt), is_integer(Value) ->
    Value;
validate_term_cause_mapping(Opt, Value) ->
    throw({error, {termination_cause_mapping, {Opt, Value}}}).

validate_term_cause_mapping(Opts) ->
    ergw_aaa_config:validate_options(
      fun validate_term_cause_mapping/2, Opts, ?DEFAULT_TERMINATION_CAUSE_MAP, map).
```

Note: this change termination_cause_mapping ot a **map**, code using id needs to be adjusted.

**Review by RoadRunnr — CHANGES_REQUESTED** [review]

File: include/ergw_aaa.hrl (hunk 2 above)
[the remark this answers is not in the document]
Comment (vkatsuba) [reply include/ergw_aaa.hrl]: Done

File: src/ergw_aaa_diameter.erl [hunk 7]
```diff
@@ -150,6 +152,17 @@ validate_transport(Opt, Value) ->
...
+    ergw_aaa_config:validate_options(fun validate_option/2, Opts, ?DefaultOptions, map).
+
+validate_option(termination_cause_mapping = Opt, Value) when is_list(Value) ->
```
[the remark this answers is not in the document]
Comment (vkatsuba) [reply src/ergw_aaa_diameter.erl]: Done

File: src/ergw_aaa_session.erl (hunk 1 above)
[the remark this answers is not in the document]
Comment (vkatsuba) [reply src/ergw_aaa_session.erl]: Done

File: src/ergw_aaa_diameter.erl (hunk 3 above)
Comment (vkatsuba) [reply src/ergw_aaa_diameter.erl]: Done

File: src/ergw_aaa_diameter.erl (hunk 4 above)
Comment (vkatsuba) [reply src/ergw_aaa_diameter.erl]: Removed. Done.

File: src/ergw_aaa_gx.erl (hunk 5 above)
Comment (vkatsuba) [reply src/ergw_aaa_gx.erl]: Thanks your example.

File: src/ergw_aaa_diameter.erl (hunk 6 above)
Comment (vkatsuba) [reply src/ergw_aaa_diameter.erl]: Thanks your example

File: include/ergw_aaa.hrl [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+    {administrative, 4},      % ?'DIAMETER_BASE_TERMINATION-CAUSE_ADMINISTRATIVE'
+    {link_broken, 5},         % ?'DIAMETER_BASE_TERMINATION-CAUSE_LINK_BROKEN'
+    {upf_failure, 5},         % ?'DIAMETER_BASE_TERMINATION-CAUSE_LINK_BROKEN'
+    {session_error, 1},       % ?'DIAMETER_BASE_TERMINATION-CAUSE_LOGOUT'
+    {inactivity_timeout, 1},  % ?'DIAMETER_BASE_TERMINATION-CAUSE_LOGOUT'
+    {path_failure, 1}         % ?'DIAMETER_BASE_TERMINATION-CAUSE_LOGOUT'
+]).
+
+%% RFC: https://tools.ietf.org/html/rfc2866#section-5.10
+-define(RADIUS_DEFAULT_TERMINATION_CAUSE_MAPPING, [
```
Comment (vkatsuba) [inline include/ergw_aaa.hrl:13]: @RoadRunnr I not sure about that this mapping is expected for radius, please take a look into it.

File: test/radius_SUITE.erl [hunk 9]
```diff
@@ -126,34 +127,13 @@ simple() ->
...
-    ?match({ok, _, _}, ergw_aaa_session:invoke(Session, #{}, interim, [])),
-    TermOpts = #{'Termination-Cause' => ?'DIAMETER_BASE_TERMINATION-CAUSE_LOGOUT'},
-    ?match({ok, _, _}, ergw_aaa_session:invoke(Session, TermOpts, stop, [])),
-
-    ?equal([{ergw_aaa_radius, started, 0}], get_session_stats()),
-
-    meck_validate(Config),
-    ok.
+simple_normal_terminate(Config) ->
+    simple(Config, #{reason => normal}).
```
[the remark this answers is not in the document]
Comment (vkatsuba) [reply test/radius_SUITE.erl]: Done.

**Comment by vkatsuba** [comment]
> Is there a reason to have a list that is turned into a map?
> As opposed to having the map directly.

There is no definite reason. But it seems to me that we can use both a list and a map.

File: src/ergw_aaa_config.erl [hunk 10]
```diff
@@ -61,6 +64,26 @@ get_application(AppId) ->
...
+radius_validate_term_cause_mapping(Opts) when is_list(Opts); is_map(Opts) ->
+    validate_options(fun radius_validate_term_cause_mapping/2, Opts, ?RADIUS_DEFAULT_TERMINATION_CAUSE_MAPPING, map);
+radius_validate_term_cause_mapping(Opts) ->
+    throw({error, {termination_cause_mapping, Opts}}).
+
+radius_validate_term_cause_mapping(Opt, Value) when is_list(Opt), is_integer(Value) ->
+    Value;
+radius_validate_term_cause_mapping(Opt, Value) ->
+    throw({error, {termination_cause_mapping, {Opt, Value}}}).
+
```
Comment (RoadRunnr) [inline src/ergw_aaa_config.erl:86]: I would prefer to have the DIAMETER part in ergw_aaa_diameter and the RADIUS part in ergw_aaa_radius. Basically keep it close to the general DIAMETER and RADIUS logic.

File: src/ergw_aaa_diameter.erl [hunk 11]
```diff
@@ -21,6 +20,7 @@
 -include_lib("diameter/include/diameter_gen_base_rfc6733.hrl").
 -include("include/diameter_3gpp_ts29_061_sgi.hrl").
 
+
```
Comment (RoadRunnr) [inline src/ergw_aaa_diameter.erl:23]: please, no unneeded  whitespace changes.

File: src/ergw_aaa_radius.erl [hunk 12]
```diff
@@ -236,8 +238,10 @@ validate_option(avp_filter, Value) when is_list(Value) ->
     lists:foldl(fun validate_avp_filters/2, #{}, Value);
 validate_option(vendor_dicts, Value) when is_list(Value) ->
     lists:map(fun validate_vendor/1, Value);
+validate_option(termination_cause_mapping, Value) ->
+    ergw_aaa_config:radius_validate_term_cause_mapping(Value);
 validate_option(Opt, Value) ->
-    throw({error, {options, {Opt, Value}}}).
+    throw({error, {options4, {Opt, Value}}}).
```
Comment (RoadRunnr) [inline src/ergw_aaa_radius.erl:244]: debug leftover?

File: src/ergw_aaa_session.erl [hunk 13]
```diff
@@ -474,3 +475,13 @@ get_handler_state(_, undefined) ->
...
+
+termination_cause_mapping(#{'Termination-Cause' := Cause} = Session, #{termination_cause_mapping := Config}) when is_atom(Cause) ->
+    case maps:get(Cause, Config, undefined) of
+        undefined ->
+            Session;
+        Res ->
+            Session#{'Termination-Cause' := Res}
+    end;
+termination_cause_mapping(Session, _) ->
+    Session.
```
Comment (RoadRunnr) [inline src/ergw_aaa_session.erl:487]:
```suggestion
termination_cause_mapping(#{'Termination-Cause' := Cause} = Session, #{termination_cause_mapping := Config})
  when is_atom(Cause), is_map_key(Cause, Config) ->
    Session#{'Termination-Cause' := maps:get(Cause, Config)};
termination_cause_mapping(Session, _) ->
    Session.
```

**Review by RoadRunnr — CHANGES_REQUESTED** [review]

**Comment by ebengt** [comment]
Looks good.

File: src/ergw_aaa_radius.erl (hunk 12 above)
Comment (vkatsuba) [reply src/ergw_aaa_radius.erl]: Yep :-), thanks.

**Review by RoadRunnr — APPROVED** [review]
LGTM, please squash the commits before merging

**Pull request merged**